# Capstone Project Segmenting and Clustering Exercise #2

# Part 1

In [1]:
import requests # library to handle requests
import pandas as pd # library for data analsysis
import numpy as np # library to handle data in a vectorized manner
import random # library for random number generation

#!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # module to convert an address into latitude and longitude values

# libraries for displaying images
from IPython.display import Image 
from IPython.core.display import HTML 


from IPython.display import display_html
import pandas as pd
import numpy as np
    
# tranforming json file into a pandas dataframe library
from pandas.io.json import json_normalize

import folium # plotting library
from bs4 import BeautifulSoup
from sklearn.cluster import KMeans
import matplotlib.cm as cm
import matplotlib.colors as colors

print('Folium installed')
print('Libraries imported.')

Folium installed
Libraries imported.


### Web Scraping and Data Wrangling

In [2]:
# read the webpage from the wiki
url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
response = requests.get(url)
#Process and convert html data
data = response.text
soup = BeautifulSoup(data,'html.parser')
wiki_table=soup.find('table')
#develop dataframe
df = pd.read_html(str(wiki_table))[0]
df.head()

Postal Code           Borough              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

In [3]:
# Drop the first column
df.drop(0,inplace=True)
#Rename the columns names
df.columns = ['PostalCode','Borough','Neighborhood']
df.head()

PostalCode           Borough                      Neighborhood
1        M2A      Not assigned                      Not assigned
2        M3A        North York                         Parkwoods
3        M4A        North York                  Victoria Village
4        M5A  Downtown Toronto         Regent Park, Harbourfront
5        M6A        North York  Lawrence Manor, Lawrence Heights

In [4]:
# Remove "Borough" with 'Not assigned' values
df2=df[df['Borough'].str.contains("Not assigned") == False].reset_index()
df2.head()

index PostalCode           Borough  \
0      2        M3A        North York   
1      3        M4A        North York   
2      4        M5A  Downtown Toronto   
3      5        M6A        North York   
4      6        M7A  Downtown Toronto   

                                  Neighborhood  
0                                    Parkwoods  
1                             Victoria Village  
2                    Regent Park, Harbourfront  
3             Lawrence Manor, Lawrence Heights  
4  Queen's Park, Ontario Provincial Government

In [5]:
df3= df2.groupby(['PostalCode', 'Borough'])['Neighborhood'].apply(', '.join).reset_index()
df3

PostalCode      Borough                                       Neighborhood
0          M1B  Scarborough                                     Malvern, Rouge
1          M1C  Scarborough             Rouge Hill, Port Union, Highland Creek
2          M1E  Scarborough                  Guildwood, Morningside, West Hill
3          M1G  Scarborough                                             Woburn
4          M1H  Scarborough                                          Cedarbrae
..         ...          ...                                                ...
98         M9N         York                                             Weston
99         M9P    Etobicoke                                          Westmount
100        M9R    Etobicoke  Kingsview Village, St. Phillips, Martin Grove ...
101        M9V    Etobicoke  South Steeles, Silverstone, Humbergate, Jamest...
102        M9W    Etobicoke                Northwest, West Humber - Clairville

[103 rows x 3 columns]

In [6]:
df3.shape

(103, 3)

We have 103 Rows of data.

# Part 2

## Get Geospatial coordinates from https://cocl.us/Geospatial_data

In [7]:
# get the data and create the dataframe
url='https://cocl.us/Geospatial_data/Geospatial_Coordinates.csv'
df_geo=pd.read_csv(url)
df_geo.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [8]:
df_geo = df_geo.rename(columns = {'Postal Code':'PostalCode'}) 
df_geo.head()

PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

## Merge the Datasets

In [9]:
df3 = pd.merge(df3, df_geo, on = 'PostalCode')
df3

PostalCode      Borough  \
0          M1B  Scarborough   
1          M1C  Scarborough   
2          M1E  Scarborough   
3          M1G  Scarborough   
4          M1H  Scarborough   
..         ...          ...   
98         M9N         York   
99         M9P    Etobicoke   
100        M9R    Etobicoke   
101        M9V    Etobicoke   
102        M9W    Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                       Malvern, Rouge  43.806686 -79.194353  
1               Rouge Hill, Port Union, Highland Creek  43.784535 -79.160497  
2                    Guildwood, Morningside, West Hill  43.763573 -79.188711  
3                                               Woburn  43.770992 -79.216917  
4                                            Cedarbrae  43.773136 -79.239476  
..                                                 ...        ...        ...  
98                                              Weston  43.706876 -79.518188  
99                                           Westmount  43.696319 -79.532242  
100  Kingsview Village, St. Phillips, Martin Grove ...  43.688905 -79.554724  
101  South Steeles, Silverstone, Humbergate, Jamest...  43.739416 -79.588437  
102                Northwest, West Humber - Clairville  43.706748 -79.594054  

[103 rows x 5 columns]

In [10]:
df3.shape

(103, 5)

We still have 103 rows but we now have added 2 columns for Latitude and Longitude

# Part 3

## Prepare the Dataframe with "Borough" that contains the word 'Toronto' 

In [11]:
#We work with only "Borough" which contain the word Toronto 
df4=df3[df3['Borough'].str.contains('Toronto')]
df5=df4.reset_index(drop=True)
df5.head()


PostalCode          Borough                    Neighborhood   Latitude  \
0        M4E     East Toronto                     The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West, Riverdale  43.679557   
2        M4L     East Toronto  India Bazaar, The Beaches West  43.668999   
3        M4M     East Toronto                 Studio District  43.659526   
4        M4N  Central Toronto                   Lawrence Park  43.728020   

   Longitude  
0 -79.293031  
1 -79.352188  
2 -79.315572  
3 -79.340923  
4 -79.388790

In [12]:
df5.shape

(40, 5)

## We now get only 40 values

We will use the Borough as Label for our clustering alogithm

In [13]:
df5['Borough'].value_counts()

Downtown Toronto    19
Central Toronto      9
West Toronto         6
East Toronto         5
Toronto/York         1
Name: Borough, dtype: int64

We need to create a column for the label, there are 5 different data types for "Borough" 

In [14]:
df5['Label']=df5['Borough'].replace(to_replace=['Downtown Toronto','Central Toronto','West Toronto','East Toronto','Toronto/York'],value=[1,2,3,4,5],inplace=False)
df5

PostalCode           Borough  \
0         M4E      East Toronto   
1         M4K      East Toronto   
2         M4L      East Toronto   
3         M4M      East Toronto   
4         M4N   Central Toronto   
5         M4P   Central Toronto   
6         M4R   Central Toronto   
7         M4S   Central Toronto   
8         M4T   Central Toronto   
9         M4V   Central Toronto   
10        M4W  Downtown Toronto   
11        M4X  Downtown Toronto   
12        M4Y  Downtown Toronto   
13        M5A  Downtown Toronto   
14        M5B  Downtown Toronto   
15        M5C  Downtown Toronto   
16        M5E  Downtown Toronto   
17        M5G  Downtown Toronto   
18        M5H  Downtown Toronto   
19        M5J  Downtown Toronto   
20        M5K  Downtown Toronto   
21        M5L  Downtown Toronto   
22        M5N   Central Toronto   
23        M5P   Central Toronto   
24        M5R   Central Toronto   
25        M5S  Downtown Toronto   
26        M5T  Downtown Toronto   
27        M5V  Downtown Toronto   
28        M5W  Downtown Toronto   
29        M5X  Downtown Toronto   
30        M6G  Downtown Toronto   
31        M6H      West Toronto   
32        M6J      West Toronto   
33        M6K      West Toronto   
34        M6N      Toronto/York   
35        M6P      West Toronto   
36        M6R      West Toronto   
37        M6S      West Toronto   
38        M7A  Downtown Toronto   
39        M7Y      East Toronto   

                                         Neighborhood   Latitude  Longitude  \
0                                         The Beaches  43.676357 -79.293031   
1                        The Danforth West, Riverdale  43.679557 -79.352188   
2                      India Bazaar, The Beaches West  43.668999 -79.315572   
3                                     Studio District  43.659526 -79.340923   
4                                       Lawrence Park  43.728020 -79.388790   
5                                    Davisville North  43.712751 -79.390197   
6                   North Toronto West, Lawrence Park  43.715383 -79.405678   
7                                          Davisville  43.704324 -79.388790   
8                         Moore Park, Summerhill East  43.689574 -79.383160   
9   Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -79.400049   
10                                           Rosedale  43.679563 -79.377529   
11                        St. James Town, Cabbagetown  43.667967 -79.367675   
12                               Church and Wellesley  43.665860 -79.383160   
13                          Regent Park, Harbourfront  43.654260 -79.360636   
14                           Garden District, Ryerson  43.657162 -79.378937   
15                                     St. James Town  43.651494 -79.375418   
16                                        Berczy Park  43.644771 -79.373306   
17                                 Central Bay Street  43.657952 -79.387383   
18                           Richmond, Adelaide, King  43.650571 -79.384568   
19  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752   
20           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576   
21                     Commerce Court, Victoria Hotel  43.648198 -79.379817   
22                                           Roselawn  43.711695 -79.416936   
23    Forest Hill North & West, Forest Hill Road Park  43.696948 -79.411307   
24                The Annex, North Midtown, Yorkville  43.672710 -79.405678   
25                     University of Toronto, Harbord  43.662696 -79.400049   
26          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049   
27  CN Tower, King and Spadina, Railway Lands, Har...  43.628947 -79.394420   
28                                     Stn A PO Boxes  43.646435 -79.374846   
29             First Canadian Place, Underground city  43.648429 -79.382280   
30                                           Christie  43.669542 -79.422564   
31                       Dufferin, Dove

Lets get the Latitude and Longitude of Toronto so we can use it for our visualization.

In [17]:
address = 'Toronto'
geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print(f'The geograpical coordinate of Toronto are {latitude}, {longitude}.')

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


In [18]:
df5['Label'] = df5.Label.astype(int)

In [20]:
kclusters=len(df5.Label.unique())
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(df5['Latitude'], df5['Longitude'], df5['Neighborhood'], df5['Label']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

We can see our 5 clusters with the Boroughs that contain Toronto in the name. 

This includes a single Borough called Toronto/York in red